ste/downloader-blob.go

```diff
@@ -110,3 +132,9 @@ func (bd *blobDownloader) GenerateDownloadFunc(jptm IJobPartTransferMgr, srcPipe
 		}
 	})
 }
+
+type dummyReader struct{}
+
+func (dummyReader) Read(p []byte) (n int, err error) {
+	return len(p), nil
+}
```

Comment:
Ooooh.  I think we can't do this quite this simply. I think you have to zero out all the bytes in p.  Why?  Because we read into pooled byte arrays, so there might be content in there from a previous usage of the array! 

(another possible solution would be to tweak multiSliceBytePooler so that it always zeros out stuff before returning it. Maybe that's safer, at the cost of an little extra CPU activity.  As I type this, I think that's probably the best idea. If we only fix it here, then one day we may get the same mistake creeping in somewhere else in the code.  And given that the cost of fixing it "at the soucre", in multiSliceBytePooler is just some CPU cycles, I think that's probably OK, because we have not had any reports from users, at all, that indicate that CPU is a constraint on perf . What do you think?)

Reply:
Doh!! My bad, this is such a good catch, it would have been a data-corrupting bug.

We've actually heard about high CPU usage concerns in the past. I think we should stick to fixing it just here and not in multiSliceBytePooler, as it's not clear what'll be the perf impact. Thoughts? @JohnRusk

Reply:
Your question prompted me to do a bit of a search. It turns out that a zero-ing-out loop, written in the right form in Go, will be automatically turned into a call to memclr, which is an optimized Go runtime routine written in assembler. I.e. write a simple loop, complier turns it into a really fast memory clear operation.

Here's a description of the loop format (which is just an obvious range loop): https://github.com/golang/go/commit/f03c9202c43e0abb130669852082117ca50aa9b1

It's pretty quick - e.g. for a slice of 256 elements, it shaves 96% off the execution time of an un-optimized loop. So I think its fast enough for us to just do it.

One note/suggestion: I think what when we do this in multi-slice byte pooler, we should get the slice we are going to return, re-slice it so that its len = its capacity, use the loop to zero it all out, and only then re-slice it to the length that the user asked for. (That last bit is already in the code, but I'm suggesting we zero out to the full cap, not just to the requested len)

Reply:
Sorry about the slow reply @zezha-msft.  

Oh, and PS, my reason for being in favour of doing this in mulitSliceBytePooler is that, if the bug can nearly slip by us both here, it can probably slip by someone else in the future.